ndigits: check for invalid bases (fix #16766)

rfourquet opened the pull request:
I made `ndigits(x, b)` with `-1 <= b <= 1` an error in all cases, even if `x == 0`, as I thought it was more consistent. There were a lot of dispatch methods `ndigits[0z][nb](x, [b])`, so it's possible I messed up one of the paths.
Also I removed the `ndigits(::BigInt, [b])` method, to handle the base check in only one place, the drawback being that we then compare `x == 0` instead of the probably more efficient `x.size == 0`.
Hopefully this is overall a simplification of the `ndigits` logic.

rfourquet committed:
ndigits: check for invalid bases (fix #16766)

tkelman commented on base/intfuncs.jl:
```diff
@@ -195,36 +198,38 @@ function ndigits0znb(n::Int, b::Int)
 
-ndigits(x::Unsigned, b::Integer) = x==0 ? 1 : ndigits0z(x,Int(b))
-ndigits(x::Unsigned)             = x==0 ? 1 : ndigits0z(x)
+ndigits(x::Integer, b::Integer) =
```
use the `function` form here

tkelman commented on test/bigint.jl:
```diff
@@ -275,10 +275,10 @@ ndigits_mismatch(n) = ndigits(n) != ndigits(BigInt(n))
+ndigits(big(rand(Int)), big(2)^rand(2:999))
 
-@test_throws DomainError ndigits(rand(big(-999:999)), rand(typemin(Int):1))
+@test_throws DomainError ndigits(big(rand(Int)), rand(-1:1))
```
lets just check all 3 cases for the base instead of doing rand for that

tkelman commented on test/intfuncs.jl:
```diff
@@ -79,10 +79,14 @@ end
+    @test ndigits(n, b) == ndigits(big(n), b)
 end
 
+@test_throws DomainError ndigits(rand(Int), rand(-1:1))
```
same here

rfourquet commented:
Thanks for reviewing @tkelman, comments addressed.

tkelman commented:
LGTM! The added comments are quite useful, thanks for those.

rfourquet commented:
I must admit that the "0z" suffix is not 100% clear for me, hence left uncommented!

StefanKarpinski commented:
If keyword arguments didn't introduce too much overhead still, we could make the signature

```jl
digits(n; base::Integer=10, min::Integer=1)
```

That way you could do `digits(n, min=0)` instead of `digits0z`.

rfourquet commented:
Do we agree on the idea that: `ndigits(x, b)` with `-1 <= b <= 1` is an error in all cases, even if `x == 0`?
Does the `z` in `ndigits0z` stands for zero? I read that as "same as ndigits but gives zero on input 0". If this is guaranteed, it can be a useful function in user-code (e.g. [here](https://github.com/JuliaLang/julia/blob/426ad102204f6cbf30905d74c2d34ef9a6ab7f8b/base/dSFMT.jl#L96)). I modified `BigInt`'s version accordingly, but I could add some tests and doc.

StefanKarpinski commented:
> Does the z in ndigits0z stands for zero? I read that as "same as ndigits but gives zero on input 0".

Yes, that's the correct interpretation.

rfourquet committed:
rename ndigits0z to ndigits0zpb

rfourquet commented:
I pushed a new commit which renames `ndigits0z` to `ndigits0zpb` and makes `ndigits0z(n, b) = ndigits(n, b, min1=true)` a very thin wrapper over `ndigits` (`min1` is a positional argument here),  which could be removed (in line with what Stefan suggested [above](https://github.com/JuliaLang/julia/pull/16841#issuecomment-225009465)). Makes sense?

tkelman commented:
Sorry this got neglected, could you rebase?

rfourquet commented:
Updated, but please give me a couple more days before merging, I would like to squash and review once more (I can't continue working on it right now).

rfourquet committed:
ndigits: add a 'pad' argument

rfourquet commented:
Rebased, and I replaced my `min1` implementation mentioned above with a `pad` positional argument, in line with what Stefan had suggested above (not documented, should probably become a keyword arg eventually). I'm fine with renaming it to `min` or with removing it altogether.

tkelman commented:
Great, LGTM. Anyone else have any thoughts?

StefanKarpinski commented:
Let's get some votes on `pad` versus `min`:

StefanKarpinski commented:
`pad`

StefanKarpinski commented:
`min`

kshyatt commented:
Needs a rebase.

rfourquet commented:
Should I wait till 0.6 is released before rebasing and hopefully merging?

StefanKarpinski commented:
Yes, I'm afraid that since this includes user-visible changes, we'll have to wait at this point, so no point in rebasing before we branch release-0.6 from master.

rfourquet commented:
I rebased now, the travis errors are apparently independant. This is good to go on my side.

tkelman commented on base/intfuncs.jl:
```diff
@@ -379,20 +397,36 @@ function ndigits0z(n::Unsigned, b::Int)
+ndigits0zpb(x::Base.BitSigned, b::Integer) = ndigits0zpb(unsigned(abs(x)), Int(b))
+ndigits0zpb(x::Base.BitUnsigned, b::Integer) = ndigits0zpb(x, Int(b))
+
+# The suffix "0z" means that the ouput is 0 on input zero (cf. #16841)
```
output

rfourquet replied:
Thanks!

stevengj commented on base/intfuncs.jl:
```diff
@@ -346,9 +348,20 @@ function ndigits0z(x::UInt128)
+
+# TODO (when keywords args are fast): rename to ndigits and make pad a keyword
+ndigits10(x::Integer, pad::Int=1) = max(pad, ndigits0z(x))
+ndigits(x::Integer) = x == 0 ? 1 : ndigits0z(x)
```
`iszero(x)` would be more efficient for BigInt, I think.

rfourquet replied:
Done, thank you.

rfourquet commented:
I will merge in a couple of days if no objections.
